### ISSUE TITLE
Use Netty reactive streams implementation

### ISSUE THREAD

**Issue by jroper** [playframework/playframework#4727]
Currently, Play's Netty backend converts Akka streams to iteratees via reactive streams, and then puts that on top of Netty.  Akka streams should be bridge directly onto Netty through a reactive streams implementation.

But of course, first, a Netty reactive streams implementation is needed, see:

https://github.com/netty/netty/issues/3902

**Comment by hepin1989** [comment]
@jroper I suggest you start it based on netty4 and @benmccann may want netty4 too.:P
typesafe stack is going to be totally reactive streaming now.

**Comment by viktorklang** [comment]
Having a thin bridge Netty<->RS would be very nice.

**Comment by hepin1989** [comment]
this should help https://github.com/netty/netty/pull/3879

**Comment by hepin1989** [comment]
one question here,for the inbound,will the netty's channel be a ```Publisher<Byte>``` or ```Publisher<HttpRequest/WebSocketFrame>```?

**Comment by jroper** [comment]
I think it might be `Publisher<T extends MessageEvent>`, with the ability to refine to `WebSocketFrame`, `HttpChunk`, etc.  Or it could be `Publisher<T>`, with the ability to map `MessageEvent` to `ByteBuffer`, for example.

**Comment by hepin1989** [comment]
I just take a quick try as 
```scala
import akka.util.ByteString
import io.netty.channel.{ ChannelHandlerContext, SimpleChannelInboundHandler, Channel }
import org.reactivestreams.{ Subscription, Subscriber, Publisher }

/**
 * Created by kerr.
 */
object ChannelStream {
  abstract class AbstractChannelSubscriber[O](channel: Channel) extends Subscriber[O] {
    private var subscription: Subscription = _
    override def onError(t: Throwable): Unit = {
      //stop
    }

    override def onSubscribe(s: Subscription): Unit = {
      subscription = s
    }

    override def onComplete(): Unit = {
      //stop
    }

    protected def requestMore: Int

    protected def transform(t: O): Any = t

    override def onNext(t: O): Unit = {
      channel.writeAndFlush(transform(t))
      //may depends on the type of O
      //for bytes
      if (channel.isWritable) {
        subscription.request(requestMore)
      }
    }
  }

  class BytesChannelSubscriber(channel: Channel) extends AbstractChannelSubscriber[ByteString](channel) {
    override protected def requestMore: Int = channel.bytesBeforeUnwritable

    override protected def transform(t: ByteString): Any = channel.write(channel.alloc().buffer(t.length).writeBytes(t.asByteBuffer))
  }

  class MessageChannelSubscriber[O](channel: Channel) extends AbstractChannelSubscriber[O](channel) {
    override protected def requestMore: Int = 1
  }

  class ChannelPublisher[I](channel: Channel) extends SimpleChannelInboundHandler[I] with Publisher[I] {
    private var demand: Long = 0
    private var subscriber: Subscriber[_ >: I] = _

    override def subscribe(s: Subscriber[_ >: I]): Unit = {
      s.onSubscribe(new Subscription {
        override def cancel(): Unit = channel.close()

        override def request(n: Long): Unit = {
          channel.pipeline().fireUserEventTriggered(Request(n))
        }
      })
      subscriber = s
    }

    override def userEventTriggered(ctx: ChannelHandlerContext, evt: scala.Any): Unit = {
      evt match {
        case Request(n) =>
          demand += n
        case _ => super.userEventTriggered(ctx, evt)
      }
    }

    override def channelRead0(ctx: ChannelHandlerContext, msg: I): Unit = {
      subscriber.onNext(msg)
      demand -= 1
    }

    override def channelReadComplete(ctx: ChannelHandlerContext): Unit = {
      if (demand > 0) {
        ctx.read()
      }
    }

    override def exceptionCaught(ctx: ChannelHandlerContext, cause: Throwable): Unit = {
      subscriber.onError(cause)
    }

    override def channelInactive(ctx: ChannelHandlerContext): Unit = {
      subscriber.onComplete()
    }
  }

  case class Request(n: Long)
}

```

**Comment by jroper** [comment]
This is fixed by #5010.

**Comment by hepin1989** [comment]
can't wait play2.5 release.
